### PR TITLE
Decouple PIM object data and simulated memory for functional simulation memory and runtime

### PR DESCRIPTION
**Pull request by deyuan**
Major PIMeval simulator upgrade for functional simulation memory and runtime.

Key designs:
* Support two ways of storing PIM object data
  - In the data holder per PIM object, as a byte array for fast functional simulation
  - In the simulated PIM memory, i.e., pimCore, for detailed memory behavior simulation such as row read/write. The simulated PIM memory is only instantiated for non-functional simulation
  - All the PIM memory allocation metadata in pimObjInfo and pimRegion are maintained with no change, for performance modeling and future extension
* Code organization
  - In pimResMgr.h, create pimDataHolder class
  - In pimResMgr.h/cpp, create APIs in pimObjInfo for accessing the data holder, syncing data between data holder and simulated memory, and supporting reference objects
  - In pimDevice.cpp, disable simulated memory initialization in functional simulation mode
  - In pimCmd.cpp, instead of getting operands and saving results from/to simulated memory, use the data holder in functional simulation commands. Add additional syncing APIs for supporting mixed functional and micro-ops usage in non-functional mode.

Testing:
- test-functional for functional simulation
- bit-serial for non-functional simulation with mixed functional and micro-ops APIs
- test-simdram for dual-contact rows
- PIMbench apps

This closes Issue #166.

### CODE REVIEW

File: libpimeval/src/pimResMgr.h [hunk 1]
```diff
@@ -71,6 +72,94 @@ class pimRegion
...
+//! Assumption: Caller gurantees correct range and indices
+class pimDataHolder
+{
+public:
+  pimDataHolder(PimDataType dataType, uint64_t numElements)
+    : m_dataType(dataType),
+      m_numElements(numElements)
+  {
+    unsigned numBitsOfDataType = pimUtils::getNumBitsOfDataType(m_dataType);
+    m_bytesPerElement = (numBitsOfDataType + 7) / 8;
```
Comment (fasiddique) [inline libpimeval/src/pimResMgr.h:86]: Will you please add some comments on this calculation.

**Comment by fasiddique** [comment]
@deyuan thank you! This is an amazing update! As I am reviewing this code, I am having difficulty understanding the **holder** concept. Will you please elaborate?

File: libpimeval/src/pimResMgr.h (hunk 1 above)
Comment (fasiddique) [inline libpimeval/src/pimResMgr.h:86]:
```suggestion
    // Calculates bytes per element, rounding up to the nearest byte
    m_bytesPerElement = (numBitsOfDataType + 7) / 8;
```

File: libpimeval/src/pimResMgr.h [hunk 2]
```diff
@@ -71,6 +72,94 @@ class pimRegion
...
+  // copy data of range [idxBegin, idxEnd) from another holder to this holder
+  void copyFromObj(pimDataHolder& src, uint64_t idxBegin = 0, uint64_t idxEnd = 0) {
+    uint64_t byteIndex = idxBegin * m_bytesPerElement;
+    uint64_t numElements = (idxEnd == 0 ? m_numElements : idxEnd - idxBegin);
+    uint64_t numBytes = numElements * m_bytesPerElement;
+    std::memcpy(m_data.data() + byteIndex, src.m_data.data() + byteIndex, numBytes);
+  }
+
+  // copy data of range [idxBegin, idxEnd) from this holder to another holder
+  void copyToObj(pimDataHolder& dest, uint64_t idxBegin = 0, uint64_t idxEnd = 0) const {
```
Comment (fasiddique) [inline libpimeval/src/pimResMgr.h:117]: Is this device to device copy?

File: libpimeval/src/pimResMgr.h (hunk 1 above)
Comment (deyuan) [reply libpimeval/src/pimResMgr.h]: Added more comments

File: libpimeval/src/pimResMgr.h (hunk 2 above)
Comment (deyuan) [reply libpimeval/src/pimResMgr.h]: Yes

**Comment by deyuan** [comment]
> @deyuan thank you! This is an amazing update! As I am reviewing this code, I am having difficulty understanding the **holder** concept. Will you please elaborate?

I updated PR description with more details. Just let me know if there are any questions.

File: libpimeval/src/pimResMgr.cpp [hunk 3]
```diff
@@ -84,6 +87,156 @@ pimObjInfo::getRegionsOfCore(PimCoreId coreId) const
...
+  m_data.setElementBits(index, bits);
+}
+
+//! @brief  Get bit representation of an element at index, with ref support
+uint64_t
+pimObjInfo::getElementBits(uint64_t index) const
+{
+  // handle reference
+  if (m_refObjId != -1) {
+    pimObjInfo &refObj = m_device->getResMgr()->getObjInfo(m_refObjId);
```
Comment (fasiddique) [inline libpimeval/src/pimResMgr.cpp:177]:
```suggestion
    pimObjInfo& refObj = m_device->getResMgr()->getObjInfo(m_refObjId);
```

File: libpimeval/src/pimResMgr.cpp [hunk 4]
```diff
@@ -84,6 +87,156 @@ pimObjInfo::getRegionsOfCore(PimCoreId coreId) const
   return regions;
 }
 
+//! @brief  Copy data from host memory to PIM object data holder, with ref support
+void
+pimObjInfo::copyFromHost(void* src, uint64_t idxBegin, uint64_t idxEnd)
+{
+  // handle reference
+  if (m_refObjId != -1) {
+    pimObjInfo &refObj = m_device->getResMgr()->getObjInfo(m_refObjId);
```
Comment (fasiddique) [inline libpimeval/src/pimResMgr.cpp:96]:
```suggestion
    pimObjInfo& refObj = m_device->getResMgr()->getObjInfo(m_refObjId);
```

File: libpimeval/src/pimResMgr.cpp [hunk 5]
```diff
@@ -84,6 +87,156 @@ pimObjInfo::getRegionsOfCore(PimCoreId coreId) const
...
+  m_data.copyToObj(destObj.m_data, idxBegin, idxEnd);
+}
+
+//! @brief  Set an element at index with bit presentation, with ref support
+void
+pimObjInfo::setElementBits(uint64_t index, uint64_t bits)
+{
+  // handle reference
+  if (m_refObjId != -1) {
+    pimObjInfo &refObj = m_device->getResMgr()->getObjInfo(m_refObjId);
```
Comment (fasiddique) [inline libpimeval/src/pimResMgr.cpp:159]:
```suggestion
    pimObjInfo& refObj = m_device->getResMgr()->getObjInfo(m_refObjId);
```

**Review by fasiddique — APPROVED** [review]
The code looks good to me except for a few changes but I think that's fine. I do have some questions but we can sync up on Friday's meeting! Please feel free to merge it!

**Pull request merged**